content/posts/2024-12-04-improve-clang-doc.md

```diff
@@ -0,0 +1,114 @@
+
+Mentors: Petr Hosek and Paul Kirth
+
+## Project Background
+
+Clang-Doc is a documentation generator developed on top of libtooling, as an
+alternative to Doxygen. Development started in 2018, and continued through 2019,
+however it has since stalled. Currently, the tool can generate HTML, YAML, and markdown but the generated output has usability issues. This GSOC project was aimed to address the pain points regarding the output of the HTML, by adding support for various C++ constructs and reworking the css of the HTML output to be more user friendly.
+
+## Work Done
+
+The original scope of the project was to improve the output of clang-doc's generation. However during testing the tool was significantly slower than expected which made developing features for the tool impossible. 
+Each compilation of the LLVM codebase was taking upwards of 10 hours on my local machine. Additionally the tool utilized a lot of memory and was prone to crashing with an out of memory error. Similar tools such as Doxygen and Hdoc ran in comparatively less time for the same codebase. This pointed to a significant bottleneck within clang-doc’s codepath when generating large scale software projects. Due to this the project scope quickly changed to improving the runtime of clang-doc so that it could run much faster. It was only during the latter half of the project did the scope change back to improving clang-doc’s generation.
+
+### Added More Test Cases to Clang-Doc test suite
+
+
+Clang-doc previously had test’s which did not test the full scope of the tool especially with regards to the HTML or Markdown output of the tool. In order to make sure the optimization experiments were not accidentally breaking the tool, it was necessary to add more end to end tests. 
+In summary I added four comprehensive tests which covered all features that we were not testing such as testing the generation for Enums, Namespace and Records for HTML and Markdown. 
+
+
+### Improve clang-doc’s performance by 1.58 times 
+
+Internally the way clang-doc works is by leveraging libtooling's ASTVisitor class to parse the declaration in each project and serializing it into an internal format, which gets deserialized later when we output the final format. The bottleneck in clang-doc lied in the way clang-doc was doing redundant work when it was visiting each declaration. 
+
+For example if we had a the following project: 
+
+```cpp
+//File: Base.cpp
+
+class Base {}
+```
+
+
+```cpp
+//File: A.cpp
+
+class A : public Base {}
+```
+
+```cpp
+//File: B.cpp
+
+class B : public Base {}
+```
+
+In this case the ASTVisitor invoke by clang-doc would visit serialized the Base class three times, once when it is parsing Base.cpp, another when its visiting A.cpp then B.cpp. This means any C++ project that heavily leverages inheritance would result in a lot of redundant work. 
```

Comment:
```suggestion
In this case, the ASTVisitor invoked by clang-doc would visit the serialized Base class three times, once when it is parsing Base.cpp, another when its visiting A.cpp then B.cpp. This means any C++ project that heavily leverages inheritance would result in a lot of redundant work. 
```

The problem isn't inheritance, per se. A major part of it is that class/type decls are usually in headers, which appear in many TUs, and none of those are deduplicated or recognized. Visiting nodes w/in a class hierarchy multiple times is unfortunate, but if it was only a problem proportional to the depth of the class hierarchy, we wouldn't have seen such a dramatic performance improvement. The basic problem is that we never checked if we'd seen an item before and did the full processing on every instance. The memoization strategy stopped us from doing redundant things, and we got a big speedup.